clock/clock.go

```diff
@@ -30,13 +30,36 @@ type PassiveClock interface {
 // needs to do arbitrary things based on time.
 type Clock interface {
 	PassiveClock
+	// After returns the channel of a new Timer.
+	// This method does not allow to free/GC the backing timer before it fires. Use
+	// NewTimer instead.
 	After(d time.Duration) <-chan time.Time
+	// NewTimer returns a new Timer.
 	NewTimer(d time.Duration) Timer
+	// Sleep sleeps for the provided duration d.
+	// Consider making the sleep interruptible by using 'select' on a context channel and a timer channel.
 	Sleep(d time.Duration)
+	// Tick returns the channel of a new Ticker.
+	// This method does not allow to free/GC the backing ticker. Use
+	// NewTicker instead.
 	Tick(d time.Duration) <-chan time.Time
 }
 
-var _ = Clock(RealClock{})
+// WithTicker allows for injecting fake or real clocks into code that
+// needs to do arbitrary things based on time.
+type WithTicker interface {
```

Comment:
I think it'd be much more convenient if there were either a) a ClockV2 interface, or b) we made a v2 version of the clock package. (and v1 can be a wrapper that just calls the v2 package.) The latter is probably most correct but it might take some mucking about with go mod files.

Reply:
I've tried `ClockV2` - https://github.com/kubernetes/utils/pull/197#discussion_r555531160

I'm not sure how to make a "v2 version of the clock package". I can make a v2 of the whole repo, but I'm not sure what you mean here, sorry.

Reply:
I think it might be worth considering a `v2` of the package rather than adding a new clock interface. A new major version shouldn't break anyone who is using go modules.

Reply:
It's unfortunate if we have to go to v2 of everything in utils. @liggitt might know if we can do some go mod magic to make clock its own module without breaking anyone, and then bump just that to v2. I think it might be possible?

@thockin might have an opinion on whether it's worth rev'ing everything under utils if it's hard to just rev the clock package.

Reply:
> @liggitt might know if we can do some go mod magic to make clock its own module without breaking anyone, and then bump just that to v2. I think it might be possible?

There's no magic. It's possible but really tedious to manage multiple published modules in a single repo (releases have to have per-module tags, etc). I think you'd get about the same effect with a `k8s.io/utils/clock/v2` package

Reply:
Yeah that actually sounds perfect, if a little bit of work to set up (e.g. we'd want the regular clock package to just call into the v2 package for as much as possible).